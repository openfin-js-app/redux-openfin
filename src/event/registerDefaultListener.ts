import {Store, Action, Middleware} from 'redux';
import { FinApplication, FinWindow } from '../GlobalTypes';
import {appEvents, windowEvents} from './constants';
import {getStateRes} from '../window/actions/handlerActionCreator';

export default function registerDefaultListener(fin: any, store: Store<any>) {

    const finApplication:FinApplication = fin.Application.getCurrentSync();
    const finWindow:FinWindow = fin.Window.getCurrentSync();
    const {dispatch} = store;

    function _dispatch_getStateRes(event:any) {
        if(event && event.type){
            if (event.type == 'maximized'){
                dispatch(getStateRes({state:'maximized'}) as any);
            }else if (event.type == 'minimized'){
                dispatch(getStateRes({state:'minimized'}) as any);
            }else if (event.type == 'restored'){
                dispatch(getStateRes({state:'normal'}) as any);
            }
        }
    }

    appEvents.forEach(async (oneEvent)=>{
        await finApplication.addListener(oneEvent.name, (event:any)=>{
            dispatch(oneEvent.actionCreator(event) as any)
        });
    });

    windowEvents.forEach(async (oneEvent)=>{
        const theName = oneEvent.name;
        if (
            theName == 'maximized'||
            theName == 'minimized'||
            theName == 'restored'
        ){
            await finWindow.addListener(theName, (event:any)=>{
                _dispatch_getStateRes(event);
                dispatch(oneEvent.actionCreator(event) as any);
            })
        }else{
            await finWindow.addListener(theName, (event:any)=>{
                dispatch(oneEvent.actionCreator(event) as any);
            })
        }
    })
}