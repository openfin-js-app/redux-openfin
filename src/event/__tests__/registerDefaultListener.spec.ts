import {Store } from 'redux';
import registerDefaultListener from '../registerDefaultListener';

jest.useFakeTimers();

describe('Event::registerDefaultListener',()=>{

    it('register by default',()=>{
        const appAddEventListener = jest.fn((name,listener)=>{
            listener({})
        });
        const winAddEventListener = jest.fn((name,listener)=>{
            if (
                name == 'maximized'||
                name == 'minimized'||
                name == 'restored'
            ){
                listener({type:name})
            }else{
                listener({})
            }
        });
        const fin = {
            Application:{
                getCurrentSync:()=>({
                    addListener:appAddEventListener,
                })
            },
            Window:{
                getCurrentSync:()=>({
                    addListener:winAddEventListener,
                })
            }
        };
        const store:Store<any> = {
            dispatch:jest.fn(),
            getState:()=>({}),
            subscribe:jest.fn(),
            replaceReducer:jest.fn(),
        } as Store<any>;

        registerDefaultListener(fin,store);

        expect(appAddEventListener).toMatchSnapshot();
        expect(winAddEventListener).toMatchSnapshot();
        expect(store.dispatch).toMatchSnapshot();

    })

})