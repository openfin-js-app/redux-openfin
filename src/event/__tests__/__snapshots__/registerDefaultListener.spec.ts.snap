// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event::registerDefaultListener register by default 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "closed",
      [Function],
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Event::registerDefaultListener register by default 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "blurred",
      [Function],
    ],
    Array [
      "bounds-changed",
      [Function],
    ],
    Array [
      "bounds-changing",
      [Function],
    ],
    Array [
      "close-requested",
      [Function],
    ],
    Array [
      "focused",
      [Function],
    ],
    Array [
      "hidden",
      [Function],
    ],
    Array [
      "maximized",
      [Function],
    ],
    Array [
      "minimized",
      [Function],
    ],
    Array [
      "restored",
      [Function],
    ],
    Array [
      "shown",
      [Function],
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Event::registerDefaultListener register by default 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "error": false,
        "payload": Object {},
        "type": "REDUX_OPENFIN::APP_EVENT::CLOSED",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {},
        "type": "REDUX_OPENFIN::WIND_EVENT::BLURRED",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {},
        "type": "REDUX_OPENFIN::WIND_EVENT::BOUNDS_CHANGED",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {},
        "type": "REDUX_OPENFIN::WIND_EVENT::BOUNDS_CHANGING",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {},
        "type": "REDUX_OPENFIN::WIND_EVENT::CLOSE_REQUESTED",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {},
        "type": "REDUX_OPENFIN::WIND_EVENT::FOCUSED",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {},
        "type": "REDUX_OPENFIN::WIND_EVENT::HIDDEN",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {
          "state": "maximized",
        },
        "type": "REDUX_OPENFIN::GET_STATE_RES",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {
          "type": "maximized",
        },
        "type": "REDUX_OPENFIN::WIND_EVENT::MAXIMIZED",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {
          "state": "minimized",
        },
        "type": "REDUX_OPENFIN::GET_STATE_RES",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {
          "type": "minimized",
        },
        "type": "REDUX_OPENFIN::WIND_EVENT::MINIMIZED",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {
          "state": "normal",
        },
        "type": "REDUX_OPENFIN::GET_STATE_RES",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {
          "type": "restored",
        },
        "type": "REDUX_OPENFIN::WIND_EVENT::RESTORED",
      },
    ],
    Array [
      Object {
        "error": false,
        "payload": Object {},
        "type": "REDUX_OPENFIN::WIND_EVENT::SHOWN",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
